sdk/core/System.ClientModel/src/Internal/ArrayBackedPropertyBag.cs

```diff
@@ -0,0 +1,308 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.ClientModel.Internal;
+
+/// <summary>
+/// A property bag which is optimized for storage of a small number of items.
+/// If the item count is less than 2, there are no allocations. Any additional items are stored in an array which will grow as needed.
+/// MUST be passed by ref only.
+/// </summary>
+internal struct ArrayBackedPropertyBag<TKey, TValue> where TKey : struct, IEquatable<TKey>
```

Comment:
This type is duplicated from Azure.Core.  Tracking resolution of this with https://github.com/Azure/azure-sdk-for-net/issues/41035